functions/passing-parameters-by-reference.md

```diff
@@ -0,0 +1,98 @@
+# Module name: Passing Parameters by Reference
+_Skeleton descriptions are typeset in italic text,_
+_so please don't remove these descriptions when editing the topic._
+
+## Overview
+
+_Provides a short natural language abstract of the module’s contents._
+_Specifies the different levels of teaching._
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational</td>
+    <td>Avoiding copies using const-reference modifiers</td>
+  </tr>
+  <tr>
+    <td>Main</td>
+    <td>Using references to modify external data</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td></td>
+  </tr>
+</table>
+
+## Motivation
+
+_Why is this important?_
+_Why do we want to learn/teach this topic?_
+
+## Topic introduction
+
+_Very brief introduction to the topic._
+
+Explain what a reference type is and how it constrasts with a value type.
+
+## Foundational: Using reference types to avoid copies
```

Comment:
I think this is entirely the wrong way of prhasing it. Reference types make the function parameter into an alias of the argument at the calling site. That is the basic thought. You need to state that before any considerations of performance optimization. 

Now you can use call-by-reference for two reasons: 1. you want to alter the argument at the calling site 2. you want to cut down on copying. That's the second level.

And then there is the consideration that if you 1. not alter the argument and 2. still pass by reference, then you should add `const`.

But I think material should be presented in that order or increasing refinement of concepts.

Reply:
For foundational knowledge, and given how early this bit creeps into teaching, it seems like a lot to expect of the student to understand aliases and the differences between "arguments at the calling site" vs the "function parameter". As an iterative process, I would expect a student to first become familiar with the syntax and a motivating reason to use it, and to later understand the deeper meaning of it, once they have the expanded knowledge of the language to support it.

Reply:
Ok, I think I see your point. But "Avoid copies" happens because the formal parameter becomes an alias of / reference to the argument in the calling code. The motivation and the workings are so close together that I see no reason not to teach/mention immediately the concept of reference. 

Pass by value: formal parameter become local variable; compare to making a new variable in the main program.

Pass by ref: format parameter becomes reference: compare to making a reference in the main program.